File: etl/src/main/scala/org/astraea/etl/CSVReaderBuilder.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License. You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.astraea.etl

import org.apache.spark.sql.types.StructType
import org.astraea.etl.DataType.StringType
import org.astraea.etl.CSVReaderBuilder._
class CSVReaderBuilder[PassedStep <: BuildStep] private (
    var deploymentModel: String,
    var userSchema: StructType,
    var sourcePath: String,
    var sinkPath: String
) {
  protected def this() = this(
    "deploymentModel",
    CSVReader
      .createSchema(Map("Type" -> StringType), Map("Type" -> StringType)),
    "sourcePath",
    "sinkPath"
  )

  protected def this(pb: CSVReaderBuilder[_]) = this(
    pb.deploymentModel,
    pb.userSchema,
    pb.sourcePath,
    pb.sinkPath
  )

  def spark(
      deploymentModel: String
  ): CSVReaderBuilder[PassedStep with SparkStep] = {
    this.deploymentModel = deploymentModel
    new CSVReaderBuilder[PassedStep with SparkStep](this)
  }

  def schema(
      userSchema: StructType
  ): CSVReaderBuilder[PassedStep with SchemaStep] = {
    this.userSchema = userSchema
    new CSVReaderBuilder[PassedStep with SchemaStep](this)
  }

  def sourcePath(
      source: String
  ): CSVReaderBuilder[PassedStep with SourceStep] = {
    this.sourcePath = source
    new CSVReaderBuilder[PassedStep with SourceStep](this)
  }

  def sinkPath(sink: String): CSVReaderBuilder[PassedStep with SinkStep] = {
    this.sinkPath = sink
    new CSVReaderBuilder[PassedStep with SinkStep](this)
  }

  def build()(implicit ev: PassedStep =:= FullReader): CSVReader = {
    CSVReader(deploymentModel, userSchema, sourcePath, sinkPath)
  }

}

object CSVReaderBuilder {
  sealed trait BuildStep
  sealed trait SparkStep extends BuildStep
  sealed trait SchemaStep extends BuildStep
  sealed trait SourceStep extends BuildStep
  sealed trait SinkStep extends BuildStep

  type FullReader = SparkStep with SchemaStep with SourceStep with SinkStep
  def builder() = new CSVReaderBuilder[BuildStep]()
}

File: common/src/main/java/org/astraea/common/csv/CsvReaderBuilder.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License. You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.astraea.common.csv;

import static java.util.Objects.requireNonNull;

import java.io.BufferedReader;
import java.io.Reader;
import java.util.Arrays;
import java.util.List;
import java.util.NoSuchElementException;
import java.util.Optional;
import java.util.stream.IntStream;
import org.astraea.common.Utils;

public class CsvReaderBuilder {
  private final Reader reader;
  private Boolean blankLine = false;

  CsvReaderBuilder(Reader source) {
    this.reader = source;
  }

  public CsvReaderBuilder blankLine(boolean allow) {
    this.blankLine = allow;
    return this;
  }

  public CsvReader build() {
    return new CsvReaderImpl(reader, blankLine);
  }

  private static class CsvReaderImpl implements CsvReader {
    private static final String SEPARATOR = ",";
    private final BufferedReader reader;
    private final boolean blankLine;
    private long currentLine = 0;
    private int genericLength = -1;
    private String[] nextLine;

    private CsvReaderImpl(Reader reader, boolean blankLine) {
      this.reader = new BufferedReader(requireNonNull(reader, "reader is null"));
      this.blankLine = blankLine;
    }

    @Override
    public boolean hasNext() {
      if (nextLine == null) {
        nextLine =
            Utils.packException(
                () ->
                    Optional.ofNullable(reader.readLine())
                        .map(line -> line.split(SEPARATOR))
                        .orElse(null));
        if (nextLine != null) currentLine++;
      }
      return nextLine != null;
    }

    /**
     * Reads the next line from the buffer and converts to a List<String>.Checking that each row is
     * of equal length is used to ensure the consistency of the csv data. If there are no more
     * inputs, throw error. Note: It should only be used in the csv body.
     *
     * @return A List<String> with each comma-separated element as a separate entry .
     */
    @Override
    public List<String> next() {
      List<String> strings = rawNext();
      if (genericLength == -1) genericLength = strings.size();
      else if (genericLength != strings.size()) {
        if (blankLine && String.join("", strings).isBlank()) {
          try {
            rawNext();
          } catch (NoSuchElementException e) {
            return strings;
          }
        }
        throw new RuntimeException(
            "The "
                + currentLine
                + " line does not meet the criteria. Each row of data should be equal in length.");
      }

      return strings;
    }

    @Override
    public List<String> rawNext() {
      if (!hasNext()) {
        throw new NoSuchElementException("There is no next line.");
      }
      try {
        return Arrays.stream(nextLine).toList();
      } finally {
        nextLine = null;
      }
    }

    @Override
    public void skip(int num) {
      if (num > 0) {
        currentLine = currentLine + num;
        IntStream.range(0, num).forEach((ignored) -> Utils.packException(reader::readLine));
        nextLine = null;
      }
    }

    @Override
    public void close() {
      Utils.close(reader);
    }
  }
}
